File: Codes in C++/Dynamic Programming/Minimum Path Sum in a Triangular grid.cpp
class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
      int m=triangle.size();
      int n=triangle[m-1].size();
      vector<int> front(n,0),cur(n,0);
      front=triangle[m-1];
      for(int i=n-2;i>=0;i--)
      {
          for(int j=i;j>=0;j--)
          {
              int down=triangle[i][j]+front[j];
              int diag=triangle[i][j]+front[j+1];
              cur[j]=min(down,diag); 
          }
          front=cur;
      }
    return front[0];
    }
};

File: Codes in Python3/Dynamic Programming/Minimum Path Sum in a Triangular grid.py
class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        m=len(triangle)
        front=[0]*m
        cur=[0]*m
        for j in range(m):
            front[j] = triangle[m - 1][j]
    
        for i in range(m-2,-1,-1):
            for j in range(i,-1,-1):
                down= triangle[i][j]+front[j]
                diag= triangle[i][j]+front[j+1]
                cur[j]=min(down,diag)
            # Create a new list for front based on the values of cur
            front = cur[:]
        return front[0]

# class Solution:
#     def minimumTotal(self, triangle: List[List[int]]) -> int:
#         m=len(triangle)
#         dp = [[0] * len(row) for row in triangle]
#         dp[m - 1] = triangle[m - 1]
#         for i in range(m-2,-1,-1):
#             for j in range(i,-1,-1):
#                 down= triangle[i][j]+dp[i+1][j]
#                 diag= triangle[i][j]+dp[i+1][j+1]
#                 dp[i][j]=min(down,diag)
#         return dp[0][0]